sdk/core/core/src/az_json_builder.c

```diff
@@ -2,220 +2,627 @@
-    {
-      json_builder->_internal.need_comma = false;
-      remaining_json = az_span_copy_u8(remaining_json, '{');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_OBJECT_END:
-    {
-      json_builder->_internal.need_comma = true;
-      remaining_json = az_span_copy_u8(remaining_json, '}');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_ARRAY_START:
-    {
-      json_builder->_internal.need_comma = false;
-      remaining_json = az_span_copy_u8(remaining_json, '[');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_ARRAY_END:
-    {
-      json_builder->_internal.need_comma = true;
-      remaining_json = az_span_copy_u8(remaining_json, ']');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_SPAN:
-    {
-      json_builder->_internal.need_comma = true;
-      return _az_json_builder_write_span(json_builder, token._internal.span);
-    }
-    default:
-    {
-      return AZ_ERROR_ARG;
-    }
+    required_size++; // For the leading comma separator.
+  }
+
+  int32_t index_of_first_escaped_char = -1;
+  required_size += _az_json_builder_escaped_length(name, &index_of_first_escaped_char);
+
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining_json, required_size);
+
+  if (json_builder->_internal.need_comma)
+  {
+    remaining_json = az_span_copy_u8(remaining_json, ',');
   }
 
+  remaining_json = az_span_copy_u8(remaining_json, '"');
+
+  // No character needed to be escaped, copy the whole string as is.
+  if (index_of_first_escaped_char == -1)
+  {
+    remaining_json = az_span_copy(remaining_json, name);
+  }
+  else
+  {
+    // Bulk copy the characters that didn't need to be escaped before dropping to the byte-by-byte
+    // encode and copy.
+    remaining_json
+        = az_span_copy(remaining_json, az_span_slice(name, 0, index_of_first_escaped_char));
+    remaining_json = _az_json_builder_escape_and_copy(
+        remaining_json, az_span_slice_to_end(name, index_of_first_escaped_char));
+  }
+
+  remaining_json = az_span_copy_u8(remaining_json, '"');
+  remaining_json = az_span_copy_u8(remaining_json, ':');
+
+  json_builder->_internal.bytes_written += required_size;
+  json_builder->_internal.need_comma = false;
+  json_builder->_internal.token_kind = AZ_JSON_TOKEN_PROPERTY_NAME;
   return AZ_OK;
 }
 
-AZ_NODISCARD static az_result az_json_builder_write_comma(az_json_builder* self)
+static AZ_NODISCARD az_result _az_json_builder_append_literal(
+    az_json_builder* json_builder,
+    az_span literal,
+    az_json_token_kind literal_kind)
 {
-  _az_PRECONDITION_NOT_NULL(self);
+  _az_PRECONDITION_NOT_NULL(json_builder);
+  _az_PRECONDITION(
+      literal_kind == AZ_JSON_TOKEN_NULL || literal_kind == AZ_JSON_TOKEN_TRUE
+      || literal_kind == AZ_JSON_TOKEN_FALSE);
+  _az_PRECONDITION_VALID_SPAN(literal, 4, false);
+  _az_PRECONDITION(az_span_size(literal) <= 5); // null, true, or false
```

Comment:
but I can write things like
"hello", "done", "1234", "    ", etc etc...

Reply:
This is specifically for JSON **literals**, i.e. true, false, null. JSON **strings**, **numbers**, etc. don't apply here. Look at the precondition above on token kind.

Reply:
yes, I can pass a valid `literal_kind` with any az_span of size <= 5.
I think you can remove the `literal` az_span and just receive the az_json_token_kind

Reply:
This is private method that is an implementation detail. I don't want to add extra switch logic overhead when it isn't needed.